source/dem/disable_particle_contact.cc

```diff
@@ -0,0 +1,335 @@
+                }
+            }
+
+          // Store the average granular temperature and solid fraction with
+          // active cell index
+          granular_temperature_average[cell->active_cell_index()] =
+            granular_temperature_cell;
+          solid_fractions[cell->active_cell_index()] = solid_fraction;
+        }
+    }
+}
+
+
+template <int dim>
+void
+DisableParticleContact<dim>::identify_mobility_status(
+  const DoFHandler<dim> &                background_dh,
+  const Particles::ParticleHandler<dim> &particle_handler,
+  MPI_Comm                               mpi_communicator)
+{
+  // Reset cell status containers
+  status_to_cell.clear();
+  status_to_cell.resize(mobility_status::n_mobility_status);
+
+  // Create dummy dofs for background dof handler
+  const FE_Q<dim>    fe(1);
+  const unsigned int dofs_per_cell = fe.dofs_per_cell;
+
+  // Get locally owned and relevant dofs
+  const IndexSet locally_owned_dofs = background_dh.locally_owned_dofs();
+  IndexSet       locally_relevant_dofs =
+    DoFTools::extract_locally_relevant_dofs(background_dh);
+
+  // Reinit all value of mobility at nodes
+  mobility_at_nodes.reinit(locally_owned_dofs,
+                           locally_relevant_dofs,
+                           mpi_communicator);
+  mobility_at_nodes = 0;
+
+  // Empty status (3) to nodes when no particle in cell
+  for (auto cell = active_ghost_cells.begin();
+       cell != active_ghost_cells.end();)
+    {
+      // Check if the cell has any particles
+      if (particle_handler.n_particles_in_cell(*cell) == 0)
+        {
+          std::vector<types::global_dof_index> local_dofs_indices(
+            dofs_per_cell);
+          (*cell)->get_dof_indices(local_dofs_indices);
+
+          // Remove empty cell from cells to reduce the number of check
+          cell = active_ghost_cells.erase(cell);
+
+          // Assign empty status to nodes
+          for (auto dof_index : local_dofs_indices)
+            {
+              mobility_at_nodes(dof_index) = mobility_status::empty;
+            }
+        }
+      else
+        {
+          ++cell;
+        }
+    }
+
+  mobility_at_nodes.update_ghost_values();
+
+  // Mobile status (2) to nodes (no overwrite of empty status) and to cell if
+  // the criteria is respected or cell has one or many empty nodes
+  // (empty neighbor)
+  for (auto cell = active_ghost_cells.begin();
+       cell != active_ghost_cells.end();)
+    {
+      // Assign mobility status to cell if has particles and
+      // - granular temperature > limit or
+      // - solid fraction of cell < limit or
+      // - is next to an empty cell
+      const unsigned int cell_id = (*cell)->active_cell_index();
+
+      std::vector<types::global_dof_index> local_dofs_indices(dofs_per_cell);
+      (*cell)->get_dof_indices(local_dofs_indices);
+
+      // Check if cell has an empty status node
+      bool has_empty_neighbor = false;
```

Comment:
has_empty_status_neighbor I guess? "Empty neighbor" does not feel like mobility status.

Reply:
Empty status or empty cell is kinda the same thing